src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1990,10 +1990,11 @@ size_t MergeTreeData::clearOldBrokenPartsFromDetachedDirectory()
 
     renamed_parts.tryRenameAll();
 
-    for (const auto & [old_name, new_name, disk] : renamed_parts.old_and_new_names)
+    for (auto & [old_name, new_name, disk] : renamed_parts.old_and_new_names)
     {
         removeDetachedPart(disk, fs::path(relative_data_path) / "detached" / new_name / "", old_name, false);
         LOG_DEBUG(log, "Removed broken detached part {} due to a timeout for broken detached parts", old_name);
+        old_name.clear();
```

Comment:
wow, this seems really easy to forget.

is there a reason we cannot just clear `renamed_parts.old_and_new_names`?
We move all of them and clear their `old_name`, and I suppose there are just loops that check if `old_name` is empty before doing anything. Why not just skip the loop?

Reply:
Good idea, though now there is one reason for that - in this method we return `renamed_parts.old_and_new_names.size();`, therefore it cannot be cleared. It can be solved by adding an additional counter for actually renamed files, then we can clear `old_and_new_names` and avoid additional loop. @tavplubix wdyt?

Reply:
or we just store the size before clearing.

Reply:
indeed

Reply:
hm, I suppose we need to leave it the way it is. Because if there is an exception in `removeDetachedPart`, then we will try to move previously successfully deleted files and will get an exception again.

Reply:
Or a `try catch`... But I am not sure if it is better.

Reply:
ah I see, that is a valid reason, nice catch! 😄

Reply:
As far as I remember, the original idea was to make code a bit more exception-safe: rename the rest of part to their old names in case of some exception in the middle of operation (but yes, the code that checks `old_name.empty()` is quite ugly, feel free to rewrite it)

Reply:
> As far as I remember, the original idea was to make code a bit more exception-safe: rename the rest of part to their old names in case of some exception in the middle of operation 

yeah, this is what I just wrote.

Reply:
I was race condition :)